sonar-batch/src/main/java/org/sonar/batch/source/ZeroCoverageSensor.java

```diff
@@ -0,0 +1,107 @@
+import org.sonar.api.batch.sensor.Sensor;
+import org.sonar.api.batch.sensor.SensorContext;
+import org.sonar.api.batch.sensor.SensorDescriptor;
+import org.sonar.api.batch.sensor.coverage.CoverageType;
+import org.sonar.api.batch.sensor.coverage.NewCoverage;
+import org.sonar.api.measures.CoreMetrics;
+import org.sonar.api.measures.Measure;
+import org.sonar.api.utils.KeyValueFormat;
+import org.sonar.batch.scan.measure.MeasureCache;
+
+import static com.google.common.collect.Iterables.concat;
+import static com.google.common.collect.Iterables.transform;
+import static com.google.common.collect.Sets.newHashSet;
+
+@Phase(name = Phase.Name.POST)
+public final class ZeroCoverageSensor implements Sensor {
+
+  private final MeasureCache measureCache;
+
+  public ZeroCoverageSensor(MeasureCache measureCache) {
+    this.measureCache = measureCache;
+  }
+
+  @Override
+  public void describe(SensorDescriptor descriptor) {
+    descriptor.name("Zero Coverage Sensor");
+  }
+
+  @Override
+  public void execute(final SensorContext context) {
+    FileSystem fs = context.fileSystem();
+    for (InputFile f : fs.inputFiles(fs.predicates().hasType(Type.MAIN))) {
+      if (!isCoverageMeasuresAlreadyDefined(f)) {
+        Measure execLines = measureCache.byMetric(f.key(), CoreMetrics.EXECUTABLE_LINES_DATA_KEY);
+        if (execLines != null) {
+          storeZeroCoverageForEachExecutableLine(context, f, execLines);
+        }
+
+      }
+    }
+  }
+
+  private void storeZeroCoverageForEachExecutableLine(final SensorContext context, InputFile f, Measure execLines) {
+    NewCoverage newCoverage = context.newCoverage().ofType(CoverageType.UNIT).onFile(f);
+    Map<Integer, String> lineMeasures = KeyValueFormat.parseIntString((String) execLines.value());
+    for (Map.Entry<Integer, String> lineMeasure : lineMeasures.entrySet()) {
+      int lineIdx = lineMeasure.getKey();
+      if (lineIdx <= f.lines()) {
+        String value = lineMeasure.getValue();
+        if (StringUtils.isNotEmpty(value) && Integer.parseInt(value) > 0) {
+          newCoverage.lineHits(lineIdx, 0);
+        }
+      }
+    }
+    newCoverage.save();
+  }
+
+  private boolean isCoverageMeasuresAlreadyDefined(InputFile f) {
+    Set<String> metricKeys = newHashSet(transform(measureCache.byComponentKey(f.key()), new Function<Measure, String>() {
+      public String apply(Measure input) {
+        return input.getMetricKey();
+      }
+    }));
+    Function<Metric, String> metricToKey = new Function<Metric, String>() {
+      public String apply(Metric input) {
+        return input.key();
+      }
+    };
+    Set<String> allCoverageMetricKeys = newHashSet(concat(transform(CoverageType.UNIT.allMetrics(), metricToKey),
+      transform(CoverageType.IT.allMetrics(), metricToKey),
+      transform(CoverageType.OVERALL.allMetrics(), metricToKey)));
+    boolean coverageDefined = !Sets.intersection(metricKeys, allCoverageMetricKeys).isEmpty();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Immediately return this expression instead of assigning it to the temporary variable "coverageDefined". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1488)